HIP-1.md

```diff
@@ -0,0 +1,146 @@
+    // how to interpret the remaining bytes.
+    function moduleMetadata() external view returns (bytes memory);
+
+    // Called by the Inbox to determine whether the provided root
+    // is valid to verify proofs against.
+    function accept(
+        bytes32 _root,
+        bytes calldata _message,
+        bytes calldata _metadata
+    ) external returns (bool);
+}
+
+interface IInbox {
+
+    // Called by relayers to deliver messages to the recipient.
+    // Verifies the merkle proof of `_message` against `_root` and
+    // calls `recipient.interchainSecurityModule().accept()` before
+    // calling `recipient.handle()`.
+    function process(
+        bytes32 _root,
+        bytes32[32] calldata _proof,
+        bytes calldata _message,
+        bytes calldata _metadata
+    ) external;
+}
+```
+
+### **Rationale**
+
+Sovereign consensus allows for a modular approach to interchain security.
+This allows the core protocol to be forward compatible, ensuring that Hyperlane can support new security models as they are developed.
+Furthermore, it allows applications to select, configure, or invent security models that offer the most appropriate tradeoffs.
+
+There were many ways to build this feature.
+We discuss a few of the design decisions made below.
+
+#### Location of merkle proof verification
+
+The ability for relayers to inject arbitrary metadata into the `ISM` presents a natural opportunity to move the merkle proof out of the `Inbox.process()` interface and into the `ISM`.
+In this model, the `Inbox` would be responsible only for replay protection.
+
+```
+interface IInterchainSecurityModule {
+    function moduleMetadata() external view returns (bytes memory);
+
+    function accept(
+        bytes32 _root,
+        bytes32[32] calldata _proof,
+        bytes calldata _message,
+        bytes calldata _metadata
+    ) external returns (bool);
+}
+
+interface IInbox {
+    function process(bytes calldata _message) external;
+}
+```
+
+This approach has the advantage of being even more modular than the proposed solution, as the `Inbox` is unopinionated about how messages are sent.
+This allows applications to create `ISMs` that do not incorporate the incremental merkle tree.
+As a simple example, one could build an `ISM` that is a multisig on individual messages, as opposed to a merkle root.
+
+The drawback is that those `ISMs` which wish to incorporate the incremental merkle tree into their security model must do that explicitly.
+We imagine most applications to prefer the censorship resistance and economies of scale that using the incremental merkle tree allows for.
+Thus, we propose leaving proof verification in the `Inbox` so that all `ISM` developers get this "for free".
+
+#### Removal of `_index` from Inbox.process()
+
+In Hyperlane v1, the height of `_root` is passed to `Inbox.process()` as `_index`.
+This is necessary because v1 validators sign the `(_root, _index)` tuple.
```

Comment:
The `_index` passed into `Inbox.process` was only ever used for [this line](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/core/contracts/Inbox.sol#L98):
```
require(_index >= _leafIndex, "!index");
```
Because the signature verification of `(_root, _index)` happened in the InboxValidatorManager.

I'm trying to convince myself that this line is superfluous. The obvious implication is if someone manages to find a message whose messageHash is `bytes32(0)`, they could process that message, but that's realistically impossible

However I'm wondering if the following could happen for a validator set whose origin chain is A (heads up, I wrote this out as I thought it through, so it's a little messy).

1. Validator set signs a checkpoint at index `(_root, 100)`, where the chain A `count()` is 101 (recall checkpoints are 0-indexed, count() is 1-indexed). This root has all "correct" leaves for indices [0,100], but at some higher index, e.g. 200, there's a fraudulent message.

2. This signed checkpoint is used to prove the fraudulent message at index 200 on chain B. `ISM.accept()` just sees that there's a signed root, so it's happy. `Inbox.process` doesn't care if `_leafIndex > _checkpointIndex` because it doesn't know the checkpoint index, so it's also happy, and it processes the fraudulent message. So far, this isn't anything bad because validators could always commit fraud, what really matters is that their fraud can be proven with the slashing conditions back on chain A
3. Now going through slashing conditions:

**Premature checkpoint** -- the signed checkpoint index is 100 and the count is 101, so [this require](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/core/contracts/validator-manager/OutboxValidatorManager.sol#L103) `require(_signedIndex >= count, "!premature");` fails. This isn't a premature checkpoint

**Fraudulent checkpoints** -- The way [`impliesDifferingLeaf`](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/core/contracts/validator-manager/OutboxValidatorManager.sol#L213) is written today, it's able to identify fraud if the fraudulent leaf is at index `i`, there's a known proof for a leaf at index `j` for the fraudulent root, and there's a known proof for the leaf at index `j` for a cached checkpoint at index `k` (this is easily accessible), where `i <= j <= k`. It's important that `i <= j` because we try to identify differences in full subtrees.

So I think we either need to pass `_leaf` into `Inbox.process` and enforce `require(_index >= _leafIndex, "!index");`, *or* we need to modify / add a slashing condition that works where `k < i` and `i = j`. Passing in `_leaf` to `Inbox.process` feels much simpler to me

Also fwiw, if we tried to pass in the revealed fraudulent proof for the fraudulent leaf at index 200 to `fraudulentCheckpoint` along with a proof against a cached checkpoint for the zero-leaf at index 200, it'd revert in a few places:
* We'd fail [this require](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/core/contracts/validator-manager/OutboxValidatorManager.sol#L157): `require(_signedIndex >= _leafIndex, "!index");` because `100 >= 200` isn't true.
* Further down, if it were possible to get there, we'd also fail [this require](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/core/contracts/validator-manager/OutboxValidatorManager.sol#L166): `require(_cachedIndex > 0 && _cachedIndex >= _leafIndex, "!cache");`. If the fraudulent leaf is at a low index like 200, it's possible over time we could get a cached index that is finally >= 200, but the fraudulent leaf index could be at something super high, like `2^256 - 1`

It's definitely possible I'm misunderstanding something tho so please check my logic

Reply:
Great write up. 

> or we need to modify / add a slashing condition that works where k < i and i = j. Passing in _leaf to Inbox.process feels much simpler to me

I'm not sure I agree, that seems like a pretty easy slashing condition to check for? If you can prove a non-empty leaf at index `> k`, that's fraud.

Remind me why including `_index` in the checkpoint is necessary again? I always forget.

Assume a fraudulent leaf at index `i`, a signed root (index unknown), and a merkle proof of non-empty leaf index `j` against the root, and `c == Mailbox.count()`.

If `j > c`, that's fraud.

If `j <= c`, and `j >= i`, fraud can be proved by caching the latest root and calling `impliesDifferingLeaf`.

If `j <= c` and `j < i`, fraud is not provable. I'd argue this is okay. If the fraudulent validator *never* reveals a proof for leaf index `>= i`, they can never be slashed, but the fraudulent message cannot be delivered. If they do, they can be slashed. This is consistent with the idea that validators can choose censor messages for indices `>= x`, but cannot censor individual messages.

I think including `_index` is unnecessary for optimistic ISMs as well, so long as you ensure that cached roots expire before the unbonding period is up.

Reply:
cc @yorhodes curious to get your thoughts

Reply:
On a second read I'm happy to add the `If j > c, that's fraud.` slashing condition. I was hoping to minimize adding another slashing condition because they're a bit harder to reason about and it means we'll have to write extra code in the watchtowers. I also think there's something nice about enforcing invariants on the Inbox side if you can-- but I understand this is only makes sense if the `_index` is always present, which it sounds like it's not.

I haven't read the code that ended up being implemented wrt v2 merkle proofs on the destination chain, so having a hard time understanding what the current thinking is / how an ISM looks today and how to bring that into this convo

Reply:
> Remind me why including _index in the checkpoint is necessary again? I always forget.

I always forget too lol - iirc it's directly to do with the various slashing conditions holding up depending on the index of the checkpoint?